barretenberg/cpp/src/barretenberg/crypto/merkle_tree/lmdb_store/lmdb_database.hpp

```diff
@@ -0,0 +1,33 @@
+#pragma once
+#include "barretenberg/crypto/merkle_tree/lmdb_store/callbacks.hpp"
+#include "barretenberg/crypto/merkle_tree/lmdb_store/lmdb_db_transaction.hpp"
+#include "barretenberg/crypto/merkle_tree/lmdb_store/lmdb_environment.hpp"
+
+namespace bb::crypto::merkle_tree {
+/**
+ * RAII wrapper atound the opening and closing of an LMDB database
+ * Contains a reference to its LMDB environment
+ */
+class LMDBDatabase {
+  public:
+    LMDBDatabase(const LMDBEnvironment& env,
+                 const LMDBDatabaseCreationTransaction& transaction,
+                 const std::string& name,
+                 bool integerKeys = false,
+                 bool reverseKeys = false,
+                 MDB_cmp_func* cmp = nullptr);
+
+    LMDBDatabase(const LMDBDatabase& other) = delete;
+    LMDBDatabase(LMDBDatabase&& other) = delete;
+    LMDBDatabase& operator=(const LMDBDatabase& other) = delete;
+    LMDBDatabase& operator=(LMDBDatabase&& other) = delete;
+
+    ~LMDBDatabase();
+
+    const MDB_dbi& underlying() const;
+
+  private:
+    MDB_dbi _dbi;
+    const LMDBEnvironment& _environment;
```

Comment:
We discussed references vs other things. In this case it looked like you share environments so this is ok. In order of preference
- unique_ptr (ownership)
- reference / shared_ptr
- raw pointer

whether to use shared_ptr or not, is a difficult copy. It's worth reading whatever you can find on the internet. At Google I think I never used shared_ptrs. Some TL;DR: with references you are forced to manage lifetime of the object at the caller (of this constructor), and are at risk of dangling references if you destruct the object too soon. With shared_ptrs you can do whatever you want, even destruct the original object, but you are then at risk of memory leaks (reference count never reaching 0 if things are wrong in the logic). Also shared_ptrs (in functions and constructors) pollute the interfaces. You'll see it in BB code.

Reply:
Ah, also, like we discussed, using references makes the structure non-assignable, which is sometimes a problem.